pypeit/core/wavecal/wvutils.py

```diff
@@ -47,9 +51,229 @@ def arc_lines_from_spec(spec, min_nsig =10.0, nonlinear_counts = 1e10):
+    positive shift means inspec2 is shifted to the right (higher pixel values) relative to inspec1. The convention for the stretch is
+    that it is float near unity that increases the size of the inspec2 relative to the original size (which is the size of inspec1)
+
+    Parameters
+    ----------
+    inspec1 : ndarray
+      Reference spectrum
+    inspec2 : ndarray
+      Spectrum for which the shift and stretch are computed such that it will match inspec1
+
+    Optional Parameters
+    -------------------
+    smooth: float, default
+      Gaussian smoothing in pixels applied to both spectra for the computations. Default is 5.0
+    percent_ceil: float, default=90.0
+      Appply a ceiling to the input spectra at the percent_ceil percentile level of the distribution of peak amplitudes.
+      This prevents extremely strong lines from completely dominating the cross-correlation, which can
+      causes the cross-correlation to have spurious noise spikes that are not the real maximum.
+    shift_mnmx: tuple of floats, default = (-0.05,0.05)
+      Range to search for the shift in the optimization about the initial cross-correlation based estimate of the shift.
+      The optimization will search the window (shift_cc + nspec*shift_mnmx[0],shift_cc + nspec*shift_mnmx[1]) where nspec
+      is the number of pixels in the spectrum
+    stretch_mnmx: tuple of floats, default = (0.97,1.03)
+      Range to search for the stretch in the optimization. The code may not work well if this range is significantly expanded
+      because the linear approximation used to transform the arc starts to break down.
+
+    Returns
+    -------
+    success: boolean
+      boolean indicating whether the optimization exited successfully indicating that the shift and stretch are reliable
+    shift: float
+      the optimal shift which was determined
+    stretch: float
+      the optimal stretch which was determined
+    cross_corr: float
+      the value of the cross-correlation coefficient at the optimal shift and stretch. This is a number between zero and unity,
+      which unity indicating a perfect match between the two spectra.
+    shift_init:
+      The initial shift determined by maximizing the cross-correlation coefficient without allowing for a stretch.
+    cross_corr_init:
+      The maximum of the initial cross-correlation coefficient determined without allowing for a stretch
+    """
+
+    nspec = inspec1.size
+
+    y1 = smooth_and_ceil(inspec1,smooth,percent_ceil)
+    y2 = smooth_and_ceil(inspec2,smooth,percent_ceil)
+
+    # Do the cross-correlation first and determine the
+    shift_cc, cc_val = xcorr_shift(y1, y2, smooth = None, percent_ceil = None, debug = debug)
+
+    bounds = [(shift_cc + nspec*shift_mnmx[0],shift_cc + nspec*shift_mnmx[1]), stretch_mnmx]
+    result = scipy.optimize.differential_evolution(zerolag_shift_stretch, args=(y1,y2), tol = 1e-4,
```

Comment:
Can you reduce `tol` to 1e-3 and still have it work
and speed things up?  Just a thought.

Reply:
It might work. I don't really understand how tol works. I chose 1e-4  becuase a stretch of 1.0 + 1e-3 gives 1 pixel per 1000 pixels of spectrum and our tolerance for matching is 2 pixels. 

I think there are several places where we can optimize here, but someone needs to spend a PR on that.